capsules/src/gpio_async.rs

```diff
@@ -25,27 +25,27 @@
 //! }
 //! ```
 
-use kernel::common::cells::OptionalCell;
+use core::cell::Cell;
 use kernel::hil;
-use kernel::ReturnCode;
-use kernel::{AppId, Callback, LegacyDriver};
+use kernel::{AppId, Callback, ErrorCode};
+use kernel::{CommandResult, Driver, ReturnCode};
 
 /// Syscall driver number.
 use crate::driver;
 pub const DRIVER_NUM: usize = driver::NUM::GpioAsync as usize;
 
 pub struct GPIOAsync<'a, Port: hil::gpio_async::Port> {
     ports: &'a [&'a Port],
-    callback: OptionalCell<Callback>,
-    interrupt_callback: OptionalCell<Callback>,
+    callback: Cell<Callback>,
+    interrupt_callback: Cell<Callback>,
 }
 
 impl<'a, Port: hil::gpio_async::Port> GPIOAsync<'a, Port> {
     pub fn new(ports: &'a [&'a Port]) -> GPIOAsync<'a, Port> {
         GPIOAsync {
-            ports: ports,
-            callback: OptionalCell::empty(),
-            interrupt_callback: OptionalCell::empty(),
+            ports,
+            callback: Cell::new(Callback::default()),
```

Comment:
Hm, this only works because `Callback: Copy`, which I think we wanted to remove (at least to keep the Tock 2.0 callback swap protections in check). I'd read up on the core notes, but it seems the PR is not up yet.

Having `Callback: Copy` makes checking against capsule swaps substantially harder and increases overhead in the options I considered. Maybe we should remove that (with the risk of having to change a lot of code)?

That is a general question though, which should've been considered prior to giving the green light on porting. :smile: I'll publish my current approach on the anti-swap checks tomorrow, which `Copy` would break...

Reply:
We can still use a `Cell` even without `Callback: Copy`. The only method I use that requires `Callback: Copy` is the `Cell::get()`, which could be replaced with `Cell::get_mut()`.

FWIW my analog_comparator to 2.0 PR used this same approach

Reply:
Yes, `get_mut` works without `Copy`, but requires a mutable reference of the `Cell`. I've identified `RefCell` to work in the case where we used to have an `OptionalCell<Callback>` in Tock 1.x, see #2282.